RecoLocalCalo/HGCalRecAlgos/plugins/alpaka/HGCalRecHitCalibrationESProducer.cc

```diff
@@ -1,105 +1,132 @@
-        if (row >= values.size())
-          throw edm::Exception(edm::errors::LogicError, "HGCalCalibrationESProducer")
-              << " Tried to copy row " << row << " to SoA, but only have " << values.size() << " values in JSON!";
-        if (!values.empty() && int(values[row].size()) != soa.size())
-          throw edm::Exception(edm::errors::LogicError, "HGCalCalibrationESProducer")
-              << " Expected " << soa.size() << " elements in Eigen vector, but got " << values[row].size() << "!";
-        for (int i = 0; i < soa.size(); i++)
-          soa(i) = values[row][i];
+        if (idx >= sfs.size()) {
+          cms::Exception ex("InvalidData");
+          ex << "The index of the thickness correction ()" << idx << ") for celltype " << celltype << " in layer"
+             << detid.layer() << "is too large for '" << fname << "'(" << sfs.size() << ")!";
+          ex.addContext("Calling hgcal::getThicknessCorrection()");
+        }
+        return sfs[idx];
       }
 
       // @short create the ESProducer product: a SoA with channel-level calibration constants
       std::optional<hgcalrechit::HGCalCalibParamHost> produce(const HGCalModuleConfigurationRcd& iRecord) {
-        auto const& moduleMap = iRecord.get(indexToken_);
+        auto const& moduleIndexer = iRecord.get(indexToken_);
+        auto const& moduleMapper = iRecord.get(mapToken_);
         edm::LogInfo("HGCalCalibrationESProducer") << "produce: filename=" << filename_.fullPath().c_str();
 
         // load dense indexing
-        const uint32_t nchans = moduleMap.getMaxDataSize();  // channel-level size
+        const uint32_t nchans = moduleIndexer.getMaxDataSize();  // channel-level size
         hgcalrechit::HGCalCalibParamHost product(nchans, cms::alpakatools::host());
 
         // load calib parameters from JSON
         std::ifstream infile(filename_.fullPath().c_str());
+        std::ifstream infileEnergy(filenameEnergy_.fullPath().c_str());
         json calib_data = json::parse(infile, nullptr, true, /*ignore_comments*/ true);
-        for (const auto& it : moduleMap.getTypecodeMap()) {  // loop over all module typecodes
-          std::string const& module = it.first;              // module typecode, e.g. "ML-F3PT-TX-0003"
+        json energy_data = json::parse(infileEnergy, nullptr, true, /*ignore_comments*/ true);
+
+        // check keys
+        const std::vector<std::string> energy_keys = {"dEdx", "SF_thickness_Si", "SF_thickness_SiPM"};
+        check_keys(energy_data, energy_keys, filenameEnergy_.fullPath());
+        const float nlayers = energy_data["dEdx"].size();  // number of absorber layers
+        if (nlayers != 47)                                 // TODO: retrieve from nlayers from Geometry
+          edm::LogError("HGCalCalibrationESProducer")
+              << "Expected 47 layers, but got " << nlayers << " in " << filenameEnergy_.fullPath();
```

Comment:
Can downstream handle the case `nlayers != 47` gracefully, or can it lead to crashes or other undefined behavior?